File: app/src/main/java/mariuszbaleczny/compass/location/LocationHelper.java
package mariuszbaleczny.compass.location;

import android.location.Location;

import mariuszbaleczny.compass.Utils;

public class LocationHelper {

    private Double latitude;
    private Double longitude;
    private boolean latitudeCorrect;
    private boolean longitudeCorrect;
    private Location location;

    public LocationHelper(Location location) {
        latitude = Double.NaN;
        longitude = Double.NaN;
        longitudeCorrect = false;
        latitudeCorrect = false;
        this.location = location;
    }

    public Location getLocation() {
        if (isCorrect() && location != null) {
            location.setLatitude(latitude);
            location.setLongitude(longitude);
            return location;
        }
        return null;
    }

    public boolean isCorrect() {
        return latitudeCorrect && longitudeCorrect;
    }

    public boolean setLatitude(Double latitude) {
        if (latitude != null) {
            latitudeCorrect = Utils.isCoordinateInRange(latitude, true);
            this.latitude = latitudeCorrect ? latitude : Double.NaN;
        }
        return latitudeCorrect;
    }

    public boolean setLongitude(Double longitude) {
        if (longitude != null) {
            longitudeCorrect = Utils.isCoordinateInRange(longitude, false);
            this.longitude = longitudeCorrect ? longitude : Double.NaN;
        }
        return longitudeCorrect;
    }

}

File: app/src/main/kotlin/mariuszbaleczny/compass/location/LocationHelper.kt
package mariuszbaleczny.compass.location

import android.location.Location
import mariuszbaleczny.compass.Utils

/**
 * Created by mariusz on 05.11.16.
 */
class LocationHelper(private val location: Location) : CoordinateValidator {

    private var latitude: Double? = Double.NaN
    private var longitude: Double? = Double.NaN
    private var latitudeCorrect: Boolean = false
    private var longitudeCorrect: Boolean = false

    override fun getLocation(): Location? {
        if (isCorrect()) {
            location.latitude = latitude!!
            location.longitude = longitude!!
            return location
        }
        return null
    }

    override fun isCorrect(): Boolean {
        return latitudeCorrect && longitudeCorrect
    }

    override fun isLatitudeCorrect(): Boolean {
        return latitude != null && Utils.isLatitudeInRange(latitude!!)
    }

    override fun isLongitudeCorrect(): Boolean {
        return longitudeCorrect
    }

    override fun setLatitude(latitude: Double?) {
        if (latitude != null) {
            latitudeCorrect = Utils.isLatitudeInRange(latitude)
            this.latitude = latitude
        } else {
            this.latitude = null
            latitudeCorrect = false
        }
    }

    override fun setLongitude(longitude: Double?) {
        if (longitude != null) {
            longitudeCorrect = Utils.isLongitudeInRange(longitude)
            this.longitude = if (longitudeCorrect) longitude else Double.NaN
        } else {
            this.longitude = null
            longitudeCorrect = false
        }
    }

    override fun setIncorrect() {
        latitudeCorrect = false
        longitudeCorrect = false
    }

    override fun setIncorrectLatitude() {
        latitudeCorrect = false
    }

    override fun setIncorrectLongitude() {
        longitudeCorrect = false
    }

}